File: packages/nars-session-core/src/surface-attachment.ts
export const NARS_SURFACE_ATTACHMENT_SCHEMA = 'narada.nars.surface_attachment.v1' as const;
export const NARS_SURFACE_ATTACHMENT_SUMMARY_SCHEMA = 'narada.nars.surface_attachment_summary.v1' as const;
export const NARS_SURFACE_ATTACHMENT_REFUSAL_SCHEMA = 'narada.nars.surface_attachment_refusal.v1' as const;

export const NARS_SURFACE_ATTACHMENT_STATES = Object.freeze([
  'requested',
  'discovering',
  'probing_health',
  'attached',
  'reconnecting',
  'stale',
  'detaching',
  'detached',
  'failed',
] as const);
export type NarsSurfaceAttachmentState = typeof NARS_SURFACE_ATTACHMENT_STATES[number];

export const NARS_SURFACE_ATTACHMENT_HEALTH_STATES = Object.freeze([
  'unknown',
  'healthy',
  'degraded',
  'unavailable',
  'stale',
] as const);
export type NarsSurfaceAttachmentHealthState = typeof NARS_SURFACE_ATTACHMENT_HEALTH_STATES[number];

export const NARS_SURFACE_ATTACHMENT_TRANSITIONS: Readonly<Record<NarsSurfaceAttachmentState, readonly NarsSurfaceAttachmentState[]>> = Object.freeze({
  requested: ['discovering', 'failed'],
  discovering: ['probing_health', 'failed'],
  probing_health: ['attached', 'reconnecting', 'stale', 'failed'],
  attached: ['reconnecting', 'stale', 'detaching', 'failed'],
  reconnecting: ['probing_health', 'attached', 'stale', 'detaching', 'failed'],
  stale: ['reconnecting', 'detaching', 'detached', 'failed'],
  detaching: ['detached', 'failed'],
  detached: [],
  failed: [],
});

export const NARS_SURFACE_ATTACHMENT_REFUSAL_CODES = Object.freeze({
  INVALID_REQUEST: 'surface_attachment_invalid_request',
  SESSION_NOT_FOUND: 'surface_attachment_session_not_found',
  SESSION_AMBIGUOUS: 'surface_attachment_session_ambiguous',
  SESSION_STALE: 'surface_attachment_session_stale',
  SESSION_HEALTH_UNAVAILABLE: 'surface_attachment_session_health_unavailable',
  ENDPOINT_UNAVAILABLE: 'surface_attachment_endpoint_unavailable',
  AUTHORITY_MISMATCH: 'surface_attachment_authority_mismatch',
  TERMINAL: 'surface_attachment_terminal',
});
export type NarsSurfaceAttachmentRefusalCode = typeof NARS_SURFACE_ATTACHMENT_REFUSAL_CODES[keyof typeof NARS_SURFACE_ATTACHMENT_REFUSAL_CODES];

export interface NarsSurfaceAttachmentCursor {
  last_sequence: number | null;
  next_sequence: number | null;
}

export interface NarsSurfaceAttachmentFailure {
  code: string;
  message: string;
  at: string;
  details?: Record<string, unknown>;
}

export interface NarsSurfaceAttachment {
  schema: typeof NARS_SURFACE_ATTACHMENT_SCHEMA;
  attachment_id: string;
  session_id: string;
  authority_runtime_id: string | null;
  surface_kind: string;
  surface_instance_id: string;
  projection_mode: string;
  view_policy: string | null;
  permission_set: string[];
  event_cursor: NarsSurfaceAttachmentCursor;
  event_endpoint: string | null;
  health_endpoint: string | null;
  attach_source: string;
  attachment_state: NarsSurfaceAttachmentState;
  health_state: NarsSurfaceAttachmentHealthState;
  created_at: string;
  updated_at: string;
  attached_at: string | null;
  detached_at: string | null;
  failure: NarsSurfaceAttachmentFailure | null;
  metadata: Record<string, unknown>;
}

export interface NarsSurfaceAttachmentCreateOptions {
  attachment_id: string;
  session_id: string;
  authority_runtime_id?: string | null;
  surface_kind: string;
  surface_instance_id: string;
  projection_mode?: string;
  view_policy?: string | null;
  permission_set?: readonly string[];
  event_cursor?: Partial<NarsSurfaceAttachmentCursor> | null;
  event_endpoint?: string | null;
  health_endpoint?: string | null;
  attach_source?: string;
  now?: string;
  metadata?: Record<string, unknown>;
}

export interface NarsSurfaceAttachmentTransitionEvidence extends Record<string, unknown> {
  now?: string;
  reason?: string;
  health_state?: NarsSurfaceAttachmentHealthState;
  event_cursor?: Partial<NarsSurfaceAttachmentCursor> | null;
  failure?: NarsSurfaceAttachmentFailure | null;
}

export interface NarsSurfaceAttachmentTransitionResult {
  changed: boolean;
  previous_record: NarsSurfaceAttachment;
  record: NarsSurfaceAttachment;
}

export interface NarsSurfaceAttachmentSummary {
  schema: typeof NARS_SURFACE_ATTACHMENT_SUMMARY_SCHEMA;
  count: number;
  attached_count: number;
  reconnecting_count: number;
  stale_count: number;
  detached_count: number;
  failed_count: number;
  health_counts: Record<NarsSurfaceAttachmentHealthState, number>;
}

export interface NarsSurfaceAttachmentRefusal {
  schema: typeof NARS_SURFACE_ATTACHMENT_REFUSAL_SCHEMA;
  code: NarsSurfaceAttachmentRefusalCode;
  message: string;
  session_id: string | null;
  surface_kind: string | null;
  candidates: readonly Record<string, unknown>[];
}

const STATE_SET = new Set<string>(NARS_SURFACE_ATTACHMENT_STATES);
const HEALTH_SET = new Set<string>(NARS_SURFACE_ATTACHMENT_HEALTH_STATES);

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value));
}

function requiredString(value: unknown, field: string): string {
  const normalized = typeof value === 'string' ? value.trim() : '';
  if (!normalized) throw new Error(`surface_attachment_${field}_required`);
  return normalized;
}

function nullableString(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

function timestamp(value: unknown, field: string, fallback: string): string {
  const candidate = typeof value === 'string' && value.trim() ? value.trim() : fallback;
  if (!Number.isFinite(Date.parse(candidate))) throw new Error(`surface_attachment_${field}_invalid`);
  return candidate;
}

function normalizeCursor(value: unknown): NarsSurfaceAttachmentCursor {
  const record = isRecord(value) ? value : {};
  const last = record.last_sequence;
  const next = record.next_sequence;
  return {
    last_sequence: last === null || last === undefined ? null : Number.isSafeInteger(last) ? Number(last) : null,
    next_sequence: next === null || next === undefined ? null : Number.isSafeInteger(next) ? Number(next) : null,
  };
}

function cloneAttachment(record: NarsSurfaceAttachment): NarsSurfaceAttachment {
  return {
    ...record,
    permission_set: [...record.permission_set],
    event_cursor: { ...record.event_cursor },
    failure: record.failure ? {
      ...record.failure,
      ...(record.failure.details ? { details: { ...record.failure.details } } : {}),
    } : null,
    metadata: { ...record.metadata },
  };
}

export function isNarsSurfaceAttachmentState(value: unknown): value is NarsSurfaceAttachmentState {
  return typeof value === 'string' && STATE_SET.has(value);
}

export function isNarsSurfaceAttachmentHealthState(value: unknown): value is NarsSurfaceAttachmentHealthState {
  return typeof value === 'string' && HEALTH_SET.has(value);
}

export function canTransitionNarsSurfaceAttachment(previousState: unknown, nextState: unknown): boolean {
  if (!isNarsSurfaceAttachmentState(previousState) || !isNarsSurfaceAttachmentState(nextState)) return false;
  if (previousState === nextState) return true;
  return NARS_SURFACE_ATTACHMENT_TRANSITIONS[previousState].includes(nextState);
}

export function assertNarsSurfaceAttachmentTransition(previousState: unknown, nextState: unknown): asserts nextState is NarsSurfaceAttachmentState {
  if (!canTransitionNarsSurfaceAttachment(previousState, nextState)) {
    throw new Error(`invalid_nars_surface_attachment_transition:${String(previousState)}:${String(nextState)}`);
  }
}

export function normalizeNarsSurfaceAttachment(value: unknown): NarsSurfaceAttachment {
  if (!isRecord(value)) throw new Error('surface_attachment_record_required');
  if (value.schema !== NARS_SURFACE_ATTACHMENT_SCHEMA) throw new Error(`surface_attachment_schema_invalid:${String(value.schema)}`);
  const now = timestamp(value.updated_at, 'updated_at', new Date(0).toISOString());
  const attachmentState = value.attachment_state;
  const healthState = value.health_state;
  if (!isNarsSurfaceAttachmentState(attachmentState)) throw new Error(`surface_attachment_state_invalid:${String(attachmentState)}`);
  if (!isNarsSurfaceAttachmentHealthState(healthState)) throw new Error(`surface_attachment_health_invalid:${String(healthState)}`);
  const permissionSet = Array.isArray(value.permission_set)
    ? value.permission_set.filter((permission): permission is string => typeof permission === 'string').map((permission) => permission.trim()).filter(Boolean)
    : [];
  const failureValue = value.failure;
  let failure: NarsSurfaceAttachmentFailure | null = null;
  if (failureValue !== null && failureValue !== undefined) {
    if (!isRecord(failureValue)) throw new Error('surface_attachment_failure_invalid');
    failure = {
      code: requiredString(failureValue.code, 'failure_code'),
      message: requiredString(failureValue.message, 'failure_message'),
      at: timestamp(failureValue.at, 'failure_at', now),
      ...(isRecord(failureValue.details) ? { details: { ...failureValue.details } } : {}),
    };
  }
  return {
    schema: NARS_SURFACE_ATTACHMENT_SCHEMA,
    attachment_id: requiredString(value.attachment_id, 'attachment_id'),
    session_id: requiredString(value.session_id, 'session_id'),
    authority_runtime_id: nullableString(value.authority_runtime_id),
    surface_kind: requiredString(value.surface_kind, 'surface_kind'),
    surface_instance_id: requiredString(value.surface_instance_id, 'surface_instance_id'),
    projection_mode: requiredString(value.projection_mode ?? 'live', 'projection_mode'),
    view_policy: nullableString(value.view_policy),
    permission_set: [...new Set(permissionSet)],
    event_cursor: normalizeCursor(value.event_cursor),
    event_endpoint: nullableString(value.event_endpoint),
    health_endpoint: nullableString(value.health_endpoint),
    attach_source: requiredString(value.attach_source ?? 'manual', 'attach_source'),
    attachment_state: attachmentState,
    health_state: healthState,
    created_at: timestamp(value.created_at, 'created_at', now),
    updated_at: now,
    attached_at: timestampOrNull(value.attached_at, 'attached_at'),
    detached_at: timestampOrNull(value.detached_at, 'detached_at'),
    failure,
    metadata: isRecord(value.metadata) ? { ...value.metadata } : {},
  };
}

function timestampOrNull(value: unknown, field: string): string | null {
  if (value === null || value === undefined) return null;
  return timestamp(value, field, new Date(0).toISOString());
}

export function createNarsSurfaceAttachment(options: NarsSurfaceAttachmentCreateOptions): NarsSurfaceAttachment {
  const now = timestamp(options.now, 'now', new Date().toISOString());
  return normalizeNarsSurfaceAttachment({
    schema: NARS_SURFACE_ATTACHMENT_SCHEMA,
    attachment_id: options.attachment_id,
    session_id: options.session_id,
    authority_runtime_id: options.authority_runtime_id ?? null,
    surface_kind: options.surface_kind,
    surface_instance_id: options.surface_instance_id,
    projection_mode: options.projection_mode ?? 'live',
    view_policy: options.view_policy ?? null,
    permission_set: options.permission_set ?? [],
    event_cursor: options.event_cursor ?? { last_sequence: null, next_sequence: null },
    event_endpoint: options.event_endpoint ?? null,
    health_endpoint: options.health_endpoint ?? null,
    attach_source: options.attach_source ?? 'manual',
    attachment_state: 'requested',
    health_state: 'unknown',
    created_at: now,
    updated_at: now,
    attached_at: null,
    detached_at: null,
    failure: null,
    metadata: options.metadata ?? {},
  });
}

export function transitionNarsSurfaceAttachment(
  attachment: NarsSurfaceAttachment,
  nextState: NarsSurfaceAttachmentState,
  evidence: NarsSurfaceAttachmentTransitionEvidence = {},
): NarsSurfaceAttachmentTransitionResult {
  const current = normalizeNarsSurfaceAttachment(attachment);
  assertNarsSurfaceAttachmentTransition(current.attachment_state, nextState);
  if (current.attachment_state === nextState) return { changed: false, previous_record: cloneAttachment(current), record: cloneAttachment(current) };
  const updatedAt = timestamp(evidence.now, 'now', new Date().toISOString());
  const nextHealth = evidence.health_state && isNarsSurfaceAttachmentHealthState(evidence.health_state)
    ? evidence.health_state
    : current.health_state;
  const nextCursor = evidence.event_cursor === undefined ? current.event_cursor : normalizeCursor(evidence.event_cursor);
  const failure = evidence.failure === null
    ? null
    : evidence.failure && isRecord(evidence.failure)
      ? normalizeFailure(evidence.failure, updatedAt)
      : nextState === 'failed' && typeof evidence.error === 'string'
        ? { code: 'surface_attachment_failed', message: evidence.error, at: updatedAt }
        : current.failure;
  const next = normalizeNarsSurfaceAttachment({
    ...current,
    attachment_state: nextState,
    health_state: nextHealth,
    event_cursor: nextCursor,
    updated_at: updatedAt,
    attached_at: nextState === 'attached' ? current.attached_at ?? updatedAt : current.attached_at,
    detached_at: nextState === 'detached' || nextState === 'failed' ? current.detached_at ?? updatedAt : current.detached_at,
    failure,
    metadata: {
      ...current.metadata,
      ...(typeof evidence.reason === 'string' && evidence.reason.trim() ? { last_transition_reason: evidence.reason.trim() } : {}),
    },
  });
  return { changed: true, previous_record: cloneAttachment(current), record: next };
}

function normalizeFailure(value: Record<string, unknown>, fallbackAt: string): NarsSurfaceAttachmentFailure {
  return {
    code: requiredString(value.code, 'failure_code'),
    message: requiredString(value.message, 'failure_message'),
    at: timestamp(value.at, 'failure_at', fallbackAt),
    ...(isRecord(value.details) ? { details: { ...value.details } } : {}),
  };
}

export function summarizeNarsSurfaceAttachments(attachments: readonly NarsSurfaceAttachment[] = []): NarsSurfaceAttachmentSummary {
  const healthCounts: Record<NarsSurfaceAttachmentHealthState, number> = {
    unknown: 0,
    healthy: 0,
    degraded: 0,
    unavailable: 0,
    stale: 0,
  };
  let attachedCount = 0;
  let reconnectingCount = 0;
  let staleCount = 0;
  let detachedCount = 0;
  let failedCount = 0;
  for (const attachment of attachments) {
    const normalized = normalizeNarsSurfaceAttachment(attachment);
    healthCounts[normalized.health_state] += 1;
    if (normalized.attachment_state === 'attached') attachedCount += 1;
    if (normalized.attachment_state === 'reconnecting') reconnectingCount += 1;
    if (normalized.attachment_state === 'stale') staleCount += 1;
    if (normalized.attachment_state === 'detached') detachedCount += 1;
    if (normalized.attachment_state === 'failed') failedCount += 1;
  }
  return {
    schema: NARS_SURFACE_ATTACHMENT_SUMMARY_SCHEMA,
    count: attachments.length,
    attached_count: attachedCount,
    reconnecting_count: reconnectingCount,
    stale_count: staleCount,
    detached_count: detachedCount,
    failed_count: failedCount,
    health_counts: healthCounts,
  };
}

export function createNarsSurfaceAttachmentRefusal({
  code,
  message,
  sessionId = null,
  surfaceKind = null,
  candidates = [],
}: {
  code: NarsSurfaceAttachmentRefusalCode;
  message: string;
  sessionId?: string | null;
  surfaceKind?: string | null;
  candidates?: readonly Record<string, unknown>[];
}): NarsSurfaceAttachmentRefusal {
  return {
    schema: NARS_SURFACE_ATTACHMENT_REFUSAL_SCHEMA,
    code,
    message: requiredString(message, 'refusal_message'),
    session_id: nullableString(sessionId),
    surface_kind: nullableString(surfaceKind),
    candidates: candidates.map((candidate) => ({ ...candidate })),
  };
}

File: packages/nars-session-core/native/src/surface_attachment.rs
//! Durable client/projection attachment records.

use crate::CoreError;
use serde_json::{json, Map, Value};
use std::fs::{self, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

pub const SCHEMA: &str = "narada.nars.surface_attachment.v1";
pub const SUMMARY_SCHEMA: &str = "narada.nars.surface_attachment_summary.v1";
pub const REGISTRY_SCHEMA: &str = "narada.nars.surface_attachment_registry.v1";
pub const REGISTRY_FILE: &str = "surface-attachments.json";

pub fn create(options: &Value, session_id: &str) -> Result<Value, CoreError> {
    let now = options
        .get("now")
        .and_then(Value::as_str)
        .filter(|v| !v.trim().is_empty())
        .unwrap_or_else(|| "")
        .to_string();
    let now = if now.is_empty() { now_iso() } else { now };
    let required = |key: &str| {
        options
            .get(key)
            .and_then(Value::as_str)
            .filter(|v| !v.trim().is_empty())
            .map(str::to_string)
            .ok_or_else(|| CoreError(format!("surface_attachment_{key}_required")))
    };
    let attachment_id = required("attachment_id")?;
    let surface_kind = required("surface_kind")?;
    let surface_instance_id = required("surface_instance_id")?;
    let mut permission_set = options
        .get("permission_set")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .filter_map(|v| v.as_str().map(str::to_string))
        .filter(|v| !v.trim().is_empty())
        .collect::<Vec<_>>();
    permission_set.sort();
    permission_set.dedup();
    Ok(json!({
        "schema": SCHEMA,
        "attachment_id": attachment_id,
        "session_id": session_id,
        "authority_runtime_id": options.get("authority_runtime_id").cloned().unwrap_or(Value::Null),
        "surface_kind": surface_kind,
        "surface_instance_id": surface_instance_id,
        "projection_mode": options.get("projection_mode").and_then(Value::as_str).unwrap_or("live"),
        "view_policy": options.get("view_policy").cloned().unwrap_or(Value::Null),
        "permission_set": permission_set,
        "event_cursor": options.get("event_cursor").cloned().unwrap_or_else(|| json!({ "last_sequence": Value::Null, "next_sequence": Value::Null })),
        "event_endpoint": options.get("event_endpoint").cloned().unwrap_or(Value::Null),
        "health_endpoint": options.get("health_endpoint").cloned().unwrap_or(Value::Null),
        "attach_source": options.get("attach_source").and_then(Value::as_str).unwrap_or("manual"),
        "attachment_state": "requested",
        "health_state": "unknown",
        "created_at": now,
        "updated_at": now,
        "attached_at": Value::Null,
        "detached_at": Value::Null,
        "failure": Value::Null,
        "metadata": options.get("metadata").cloned().unwrap_or_else(|| json!({})),
    }))
}

pub fn normalize(value: &Value) -> Result<Value, CoreError> {
    let object = value
        .as_object()
        .ok_or_else(|| CoreError("surface_attachment_record_required".into()))?;
    if object.get("schema").and_then(Value::as_str) != Some(SCHEMA) {
        return Err(CoreError("surface_attachment_schema_invalid".into()));
    }
    let required = |key: &str| {
        object
            .get(key)
            .and_then(Value::as_str)
            .filter(|v| !v.trim().is_empty())
            .map(str::to_string)
            .ok_or_else(|| CoreError(format!("surface_attachment_{key}_required")))
    };
    let state = required("attachment_state")?;
    let health = required("health_state")?;
    if !is_state(&state) {
        return Err(CoreError(format!(
            "surface_attachment_state_invalid:{state}"
        )));
    }
    if !is_health(&health) {
        return Err(CoreError(format!(
            "surface_attachment_health_invalid:{health}"
        )));
    }
    let mut normalized = value.clone();
    normalized["schema"] = json!(SCHEMA);
    normalized["attachment_id"] = json!(required("attachment_id")?);
    normalized["session_id"] = json!(required("session_id")?);
    normalized["surface_kind"] = json!(required("surface_kind")?);
    normalized["surface_instance_id"] = json!(required("surface_instance_id")?);
    normalized["permission_set"] = json!(object
        .get("permission_set")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default());
    normalized["event_cursor"] = object
        .get("event_cursor")
        .cloned()
        .unwrap_or_else(|| json!({ "last_sequence": Value::Null, "next_sequence": Value::Null }));
    normalized["metadata"] = object.get("metadata").cloned().unwrap_or_else(|| json!({}));
    Ok(normalized)
}

pub fn can_transition(previous: &str, next: &str) -> bool {
    previous == next
        || matches!(
            (previous, next),
            ("requested", "discovering" | "failed")
                | ("discovering", "probing_health" | "failed")
                | (
                    "probing_health",
                    "attached" | "reconnecting" | "stale" | "failed"
                )
                | (
                    "attached",
                    "reconnecting" | "stale" | "detaching" | "failed"
                )
                | (
                    "reconnecting",
                    "probing_health" | "attached" | "stale" | "detaching" | "failed"
                )
                | (
                    "stale",
                    "reconnecting" | "detaching" | "detached" | "failed"
                )
                | ("detaching", "detached" | "failed")
        )
}

pub fn transition(value: &Value, next: &str, evidence: &Value) -> Result<Value, CoreError> {
    let current = normalize(value)?;
    let previous = current
        .get("attachment_state")
        .and_then(Value::as_str)
        .unwrap_or("requested");
    if !can_transition(previous, next) {
        return Err(CoreError(format!(
            "invalid_nars_surface_attachment_transition:{previous}:{next}"
        )));
    }
    if previous == next {
        return Ok(current);
    }
    let now = evidence
        .get("now")
        .and_then(Value::as_str)
        .filter(|v| !v.trim().is_empty())
        .unwrap_or_else(|| "")
        .to_string();
    let now = if now.is_empty() { now_iso() } else { now };
    let mut next_value = current.clone();
    next_value["attachment_state"] = json!(next);
    next_value["updated_at"] = json!(now);
    if next == "attached"
        && current
            .get("attached_at")
            .map(Value::is_null)
            .unwrap_or(true)
    {
        next_value["attached_at"] = json!(now);
    }
    if next == "detached" || next == "failed" {
        if current
            .get("detached_at")
            .map(Value::is_null)
            .unwrap_or(true)
        {
            next_value["detached_at"] = json!(now);
        }
    }
    if let Some(health) = evidence
        .get("health_state")
        .and_then(Value::as_str)
        .filter(|v| is_health(v))
    {
        next_value["health_state"] = json!(health);
    }
    if let Some(cursor) = evidence.get("event_cursor") {
        next_value["event_cursor"] = cursor.clone();
    }
    if let Some(failure) = evidence.get("failure") {
        next_value["failure"] = failure.clone();
    } else if next == "failed" {
        next_value["failure"] = json!({ "code": "surface_attachment_failed", "message": evidence.get("error").and_then(Value::as_str).unwrap_or("surface attachment failed"), "at": now });
    }
    if let Some(reason) = evidence
        .get("reason")
        .and_then(Value::as_str)
        .filter(|v| !v.trim().is_empty())
    {
        next_value["metadata"]["last_transition_reason"] = json!(reason);
    }
    Ok(next_value)
}

pub fn summary(attachments: &[Value]) -> Value {
    let mut health_counts = Map::new();
    for health in ["unknown", "healthy", "degraded", "unavailable", "stale"] {
        health_counts.insert(health.into(), json!(0));
    }
    let mut attached = 0;
    let mut reconnecting = 0;
    let mut stale = 0;
    let mut detached = 0;
    let mut failed = 0;
    for attachment in attachments {
        if let Some(state) = attachment.get("attachment_state").and_then(Value::as_str) {
            match state {
                "attached" => attached += 1,
                "reconnecting" => reconnecting += 1,
                "stale" => stale += 1,
                "detached" => detached += 1,
                "failed" => failed += 1,
                _ => {}
            }
        }
        if let Some(health) = attachment.get("health_state").and_then(Value::as_str) {
            if let Some(counter) = health_counts.get_mut(health) {
                *counter = json!(counter.as_u64().unwrap_or(0) + 1);
            }
        }
    }
    json!({ "schema": SUMMARY_SCHEMA, "count": attachments.len(), "attached_count": attached, "reconnecting_count": reconnecting, "stale_count": stale, "detached_count": detached, "failed_count": failed, "health_counts": health_counts })
}

pub fn registry_path(session_path: &Path) -> Result<PathBuf, CoreError> {
    Ok(session_path
        .parent()
        .ok_or_else(|| CoreError("surface_attachment_session_path_required".into()))?
        .join(REGISTRY_FILE))
}
pub fn read_registry(session_path: Option<&Path>, session_id: &str) -> Result<Value, CoreError> {
    let Some(session_path) = session_path else {
        return Ok(
            json!({ "schema": REGISTRY_SCHEMA, "session_id": session_id, "generated_at": now_iso(), "attachments": [] }),
        );
    };
    let path = registry_path(session_path)?;
    if !path.exists() {
        return Ok(
            json!({ "schema": REGISTRY_SCHEMA, "session_id": session_id, "generated_at": now_iso(), "attachments": [] }),
        );
    }
    let parsed: Value = serde_json::from_str(
        &fs::read_to_string(&path)
            .map_err(|_| CoreError("surface_attachment_registry_corrupt".into()))?,
    )
    .map_err(|_| CoreError("surface_attachment_registry_corrupt".into()))?;
    if parsed.get("schema").and_then(Value::as_str) != Some(REGISTRY_SCHEMA) {
        return Err(CoreError(
            "surface_attachment_registry_schema_invalid".into(),
        ));
    }
    if parsed.get("session_id").and_then(Value::as_str) != Some(session_id) {
        return Err(CoreError(
            "surface_attachment_registry_session_mismatch".into(),
        ));
    }
    let mut result = parsed;
    if let Some(items) = result.get_mut("attachments").and_then(Value::as_array_mut) {
        for item in items.iter_mut() {
            *item = normalize(item)?;
        }
    } else {
        return Err(CoreError(
            "surface_attachment_registry_attachments_invalid".into(),
        ));
    }
    Ok(result)
}

pub fn write_registry(session_path: &Path, registry: &Value) -> Result<Value, CoreError> {
    let path = registry_path(session_path)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .map_err(|_| CoreError("surface_attachment_registry_directory_failed".into()))?;
    }
    write_atomic(&path, registry)?;
    Ok(registry.clone())
}
pub fn register(
    session_path: Option<&Path>,
    session_id: &str,
    options: &Value,
) -> Result<Value, CoreError> {
    let path =
        session_path.ok_or_else(|| CoreError("surface_attachment_session_path_required".into()))?;
    let mut registry = read_registry(Some(path), session_id)?;
    let attachment = if options.get("schema").is_some() {
        normalize(options)?
    } else {
        create(options, session_id)?
    };
    if attachment.get("session_id").and_then(Value::as_str) != Some(session_id) {
        return Err(CoreError("surface_attachment_session_mismatch".into()));
    }
    let mut items = registry
        .get("attachments")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let id = attachment
        .get("attachment_id")
        .and_then(Value::as_str)
        .unwrap_or_default();
    items.retain(|item| item.get("attachment_id").and_then(Value::as_str) != Some(id));
    items.push(attachment.clone());
    registry["attachments"] = Value::Array(items);
    registry["generated_at"] = json!(now_iso());
    write_registry(path, &registry)?;
    Ok(attachment)
}
pub fn transition_in_registry(
    session_path: Option<&Path>,
    session_id: &str,
    attachment_id: &str,
    next: &str,
    evidence: &Value,
) -> Result<Value, CoreError> {
    let path =
        session_path.ok_or_else(|| CoreError("surface_attachment_session_path_required".into()))?;
    let mut registry = read_registry(Some(path), session_id)?;
    let items = registry
        .get_mut("attachments")
        .and_then(Value::as_array_mut)
        .ok_or_else(|| CoreError("surface_attachment_registry_attachments_invalid".into()))?;
    let index = items
        .iter()
        .position(|item| item.get("attachment_id").and_then(Value::as_str) == Some(attachment_id))
        .ok_or_else(|| CoreError("surface_attachment_not_found".into()))?;
    let previous = items[index].clone();
    let next_value = transition(&previous, next, evidence)?;
    let changed = next_value != previous;
    if changed {
        items[index] = next_value.clone();
        registry["generated_at"] = json!(evidence
            .get("now")
            .and_then(Value::as_str)
            .unwrap_or_else(|| ""));
        if registry["generated_at"]
            .as_str()
            .unwrap_or_default()
            .is_empty()
        {
            registry["generated_at"] = json!(now_iso());
        }
        write_registry(path, &registry)?;
    }
    Ok(json!({ "changed": changed, "previous_record": previous, "record": next_value }))
}
pub fn list(session_path: Option<&Path>, session_id: &str) -> Result<Vec<Value>, CoreError> {
    Ok(read_registry(session_path, session_id)?
        .get("attachments")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default())
}

fn is_state(value: &str) -> bool {
    [
        "requested",
        "discovering",
        "probing_health",
        "attached",
        "reconnecting",
        "stale",
        "detaching",
        "detached",
        "failed",
    ]
    .contains(&value)
}
fn is_health(value: &str) -> bool {
    ["unknown", "healthy", "degraded", "unavailable", "stale"].contains(&value)
}
fn write_atomic(path: &Path, value: &Value) -> Result<(), CoreError> {
    let temp = path.with_extension(format!(
        "tmp-{}-{}",
        std::process::id(),
        Uuid::new_v4().simple()
    ));
    let mut file = File::create(&temp)
        .map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))?;
    serde_json::to_writer_pretty(&mut file, value)
        .map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))?;
    file.write_all(b"\n")
        .map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))?;
    file.flush()
        .map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))?;
    drop(file);
    if path.exists() {
        fs::remove_file(path)
            .map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))?;
    }
    fs::rename(temp, path).map_err(|_| CoreError("surface_attachment_registry_write_failed".into()))
}
fn now_iso() -> String {
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64;
    let seconds = millis.div_euclid(1_000);
    let days = seconds.div_euclid(86_400);
    let sod = seconds.rem_euclid(86_400);
    let (year, month, day) = civil_from_days(days);
    format!(
        "{year:04}-{month:02}-{day:02}T{:02}:{:02}:{:02}.{:03}Z",
        sod / 3_600,
        (sod % 3_600) / 60,
        sod % 60,
        millis.rem_euclid(1_000)
    )
}
fn civil_from_days(days: i64) -> (i64, i64, i64) {
    let z = days + 719468;
    let era = if z >= 0 { z } else { z - 146096 }.div_euclid(146097);
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096).div_euclid(365);
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2).div_euclid(153);
    let day = doy - (153 * mp + 2).div_euclid(5) + 1;
    let month = mp + if mp < 10 { 3 } else { -9 };
    let year = y + if month <= 2 { 1 } else { 0 };
    (year, month, day)
}
